### PR TITLE
FEATURE: Introduce findAllIterator() method

### PR DESCRIPTION
**Pull request by daniellienert**
> If the amount of the stored data increases, receiving all objects using a ``findAll()`` may
> consume a lot more memory than available. In this cases, you should use the ``findAllIterator()``.
> This method returns an ``IterableResult``over which you can iterate, getting only one object at a time.

As this pattern is used a lot in Neos repositories, it would make sense to move it to the base repository class.

### CODE REVIEW

**Comment by kdambekalns** [comment]
This `Repository` is persistence-layer agnostic, so the new stuff must go into the Doctrine-specific `Repository`…

**Comment by kdambekalns** [comment]
Haha, what @kitsunet just wrote. That being said, I'd be fine with getting rid of non-Doctrine persistence for Flow 5.0 😎

**Comment by kitsunet** [comment]
I am not sure of the implications. I like the idea of having hte option of non doctrine persistence. but \o/ neos won't work with anyting else at the moment anyway.

**Comment by daniellienert** [comment]
Interesting - I did't know that this one exists. But that would mean, that e.g. `NodeDataRepository` inherits the wrong `Neos\Flow\Persistence\Repository` atm?

**Comment by albe** [comment]
Regarding removing the non-doctrine persistence:
https://twitter.com/ih8nickfinding/status/912601641068752899

**Comment by daniellienert** [comment]
@kitsunet - I moved the methods. Would you have another look?

File: Neos.Flow/Classes/Persistence/Doctrine/Repository.php [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Persistence/Doctrine/Repository.php]: We don't do empty lines before `namespace`.

File: Neos.Flow/Classes/Persistence/Repository.php [hunk 2]
```diff
@@ -27,6 +28,15 @@
...
     protected $persistenceManager;
 
+    /**
+     * Doctrine's Entity Manager. Note that "ObjectManager" is the name of the related
+     * interface ...
+     *
+     * @Flow\Inject
+     * @var ObjectManager
+     */
+    protected $entityManager;
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Persistence/Repository.php]: This is not needed (same for the related use statement)

File: Neos.Flow/Documentation/TheDefinitiveGuide/PartIII/Persistence.rst [hunk 3]
```diff
@@ -196,6 +196,7 @@ From Evans, the rules we need to enforce include:
 * When a change to any object within the Aggregate boundary is committed, all invariants
   of the whole Aggregate must be satisfied.
 
+
```
Comment (kdambekalns) [inline Neos.Flow/Documentation/TheDefinitiveGuide/PartIII/Persistence.rst]: A specific reason for this empty line?

File: Neos.Flow/Tests/Functional/Persistence/Fixtures/SubSubEntityRepository.php [hunk 4]
```diff
@@ -11,7 +11,7 @@
  * source code.
  */
 
-use Neos\Flow\Persistence\Repository;
+use Neos\Flow\Persistence\Doctrine\Repository;
```
Comment (kdambekalns) [inline Neos.Flow/Tests/Functional/Persistence/Fixtures/SubSubEntityRepository.php]: Why? Is the fixture specific for a test related to the Doctrine repository?

File: Neos.Flow/Tests/Functional/Persistence/Fixtures/SuperEntityRepository.php [hunk 5]
```diff
@@ -11,7 +11,7 @@
  * source code.
  */
 
-use Neos\Flow\Persistence\Repository;
+use Neos\Flow\Persistence\Doctrine\Repository;
```
Comment (kdambekalns) [inline Neos.Flow/Tests/Functional/Persistence/Fixtures/SuperEntityRepository.php]: Why? Is the fixture specific for a test related to the Doctrine repository?

File: Neos.Flow/Classes/Persistence/Doctrine/Repository.php [hunk 6]
```diff
@@ -109,6 +129,45 @@ public function findAll()
...
+            yield $object;
+            if ($callback !== null) {
+                call_user_func($callback, $iteration, $object);
+            }
+
+            $iteration++;
+        }
+    }
+
+
```
Comment (kdambekalns) [inline Neos.Flow/Classes/Persistence/Doctrine/Repository.php]: One empty line between methods.

File: Neos.Flow/Tests/Functional/Persistence/Fixtures/SubSubEntityRepository.php (hunk 4 above)
Comment (daniellienert) [reply Neos.Flow/Tests/Functional/Persistence/Fixtures/SubSubEntityRepository.php]: I think so. As they are only used in the `Neos\Flow\Tests\Functional\Persistence\**Doctrine**\RepositoryTest` they should be of kind doctrine repository.

File: Neos.Flow/Tests/Functional/Persistence/Fixtures/SuperEntityRepository.php (hunk 5 above)
Comment (daniellienert) [reply Neos.Flow/Tests/Functional/Persistence/Fixtures/SuperEntityRepository.php]: Same here

File: Neos.Flow/Tests/Functional/Persistence/Fixtures/SubSubEntityRepository.php (hunk 4 above)
Comment (kdambekalns) [reply Neos.Flow/Tests/Functional/Persistence/Fixtures/SubSubEntityRepository.php]: 👍

File: Neos.Flow/Tests/Functional/Persistence/Fixtures/SuperEntityRepository.php (hunk 5 above)
Comment (kdambekalns) [reply Neos.Flow/Tests/Functional/Persistence/Fixtures/SuperEntityRepository.php]: 👍

**Pull request merged**